add reactjs for single support form

tasawernawaz opened the pull request:
[LEARNER-2804](https://openedx.atlassian.net/browse/LEARNER-2804)

React JS added in this PR for uploading files, validation and submission of support form. In this PR we are using test Zendesk account for uploading files and creating tickets.

After this PR merged we need to work on following tickets to integrate with edX Zendesk account.
- https://openedx.atlassian.net/browse/LEARNER-2736
- https://openedx.atlassian.net/browse/LEARNER-2735
 
For browser test we have a separate ticket.
 - https://openedx.atlassian.net/browse/LEARNER-2581

**Sandbox Link:** https://tasawernawaz.sandbox.edx.org/support/contact_us

tasawernawaz commented:
@andy-armstrong Please review this PR.

andy-armstrong commented:
@tasawernawaz I recommend asking @arizzitano to review this as she's our React guru.

andy-armstrong commented:
@tasawernawaz It would be good to build this page using Bootstrap, if possible. I guess the styling went in an earlier PR.

andy-armstrong commented:
@tasawernawaz Is the idea that this page replaces the current support modal? Do you plan to also take on the work to merge the "Help" and "Support" buttons as described here?

https://openedx.atlassian.net/browse/LEARNER-625

tasawernawaz commented:
@andy-armstrong Bootstrap is already added on this page, and I used bootstrap to build this page.

tasawernawaz commented:
@andy-armstrong Yes once this page will fully functional it eventually replace all places where users can interact.

andy-armstrong commented:
@tasawernawaz The page is loading the v1 Sass rules, not the Bootstrap version. You need to pass ``uses_bootstrap: True`` as context to the Mako template. It looks like the Bootstrap CSS is also being loaded but that's not very clean and is risky due to CSS rule collisions.

![image](https://user-images.githubusercontent.com/5985072/31622431-6504235e-b26a-11e7-9fe9-e33f4c02267a.png)

andy-armstrong commented:
@tasawernawaz It also means you aren't pulling in the edx-bootstrap Sass rules so you aren't getting the edx.org theme applied.

andy-armstrong commented:
@tasawernawaz Actually, it might be that you are passing ``use_bootstrap: True`` but that you are missing this line in your Mako template to choose the Bootstrap CSS file:

https://github.com/edx/edx-platform/blob/master/lms/templates/ux/reference/bootstrap/course-skeleton.html#L3

```
## Override the default styles_version to use Bootstrap
<%! main_css = "css/bootstrap/lms-main.css" %>
```

tasawernawaz commented:
@andy-armstrong Thanks for the explanations. I will update code according to your suggestions. 

@arizzitano Please do a review of react JS code, once you have time.

schenedx commented on lms/djangoapps/support/static/support/jsx/errors_list.jsx:
```diff
@@ -0,0 +1,36 @@
+
+    render() {
+        return this.props.errorList.length > 0 &&
+                <div className="col-sm-12">
```
The indentation here looks weird to me. What is the desired indentation here?

tasawernawaz replied:
it should be 4 spaces, I will update,

schenedx commented on lms/djangoapps/support/static/support/jsx/errors_list.jsx:
```diff
@@ -0,0 +1,36 @@
+                        <ul>
+                            {this.props.errorList.map(error =>
+                                <li>{error}</li>
+                            )
```
Maybe collapse this line and the next into a single line?

tasawernawaz replied:
I will refactor this

schenedx commented on lms/djangoapps/support/static/support/jsx/file_upload.jsx:
```diff
@@ -0,0 +1,56 @@
+        e.preventDefault();
+        let fileToken = e.target.id,
+            fileRow = $(e.target).closest('.row'),
+            url = 'https://arbisoft.zendesk.com/api/v2/uploads/' + fileToken + '.json',
```
This URL will be updated in another PR, correct?

schenedx commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    uploadFile(e) {
+        let url = 'https://arbisoft.zendesk.com/api/v2/uploads.json?filename=',
```
You should make sure the hostname is configurable so it works across different edx-platform environments (stage vs prod)

schenedx commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,295 @@
+            $this = this;
+
+        const file = e.target.files[0],
+            accessToken = 'd6ed06821334b6584dd9607d04007c281007324ed07e087879c9c44835c684da',
```
I believe this token should not be hard coded in javascript. Please find another place so this token is configurable?

tasawernawaz commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,295 @@
+// access token
+// custom fields ids
+// https://openedx.atlassian.net/browse/LEARNER-2736
+//https://openedx.atlassian.net/browse/LEARNER-2735
```
@schenedx I have added a comment that in the follow up story I will make hostname, accesstoken and custom field ids configurable.

tasawernawaz commented:
@andy-armstrong I am unable to load bootstrap by passing `use_bootstrap: True`, and if I add 
```
## Override the default styles_version to use Bootstrap
<%! main_css = "css/bootstrap/lms-main.css" %>
```
header and footer design also becomes weird.

In previous PR bootstrap was added explicitly by in `head_extra` and was working fine. See https://github.com/edx/edx-platform/blob/master/lms/templates/support/contact_us.html#L17

tasawernawaz commented:
jenkins run quality

MichaelRoytman commented on lms/djangoapps/support/static/support/jsx/errors_list.jsx:
```diff
@@ -0,0 +1,34 @@
+/* global gettext */
+/* eslint react/no-array-index-key: 0 */
```
Why disable this rule? You should add a key for your list items to ensure that React can detect only the changed elements. This [page](indentation) is a good place to read about keys and best practices for specifying keys (e.g. don't use the array index, etc.).

tasawernawaz replied:
I disabled this because we don't have any keys to add in different errors. Secondly I can't access the link you provided. https://github.com/edx/edx-platform/pull/indentation

MichaelRoytman commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,317 @@
+      $this = this,
+      file = e.target.files[0],
+      accessToken = 'd6ed06821334b6584dd9607d04007c281007324ed07e087879c9c44835c684da',
+      maxFileSize = 5000000,  // 5mb is max limit
```
There's a linting error here. I believe you don't need two spaces between the comma and the start of the comment. I would reduce it to just one.

tasawernawaz commented on lms/djangoapps/support/static/support/jsx/file_upload.jsx:
```diff
@@ -0,0 +1,53 @@
+    const fileToken = e.target.id;
+    const fileRow = $(e.target).closest('.row');
+    const url = `https://arbisoft.zendesk.com/api/v2/uploads/${fileToken}.json`;
+    const accessToken = 'd6ed06821334b6584dd9607d04007c281007324ed07e087879c9c44835c684da';
```
Yes this will passed through settings in separate ticket. https://openedx.atlassian.net/browse/LEARNER-2736

tasawernawaz commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,317 @@
+      errorList = [],
+      $this = this,
+      file = e.target.files[0],
+      accessToken = 'd6ed06821334b6584dd9607d04007c281007324ed07e087879c9c44835c684da',
```
Yes this will passed through settings in separate ticket. https://openedx.atlassian.net/browse/LEARNER-2736

tasawernawaz commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,317 @@
+      $userInfo = $('.user-info'),
+      request = new XMLHttpRequest(),
+      $course = $('#course'),
+      accessToken = 'd6ed06821334b6584dd9607d04007c281007324ed07e087879c9c44835c684da',
```
Yes this will passed through settings in separate ticket. https://openedx.atlassian.net/browse/LEARNER-2736

tasawernawaz commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,317 @@
+      course = $course.val();
+    }
+
+    data.custom_fields = [{
```
We have some custom fields on Zendesk. We need to pass values to those fields along with the id provided by Zendesk. The hard coded id will be removed in  https://openedx.atlassian.net/browse/LEARNER-2736.

tasawernawaz commented:
@MichaelRoytman Thanks for the detailed review.

I gone through all your comments, suggestions and questions. I have updated code according to your suggestions and answer to your questions. Please do another review.

arizzitano commented:
Looks good! Thanks @tasawernawaz for going the extra mile to introduce React here, cheers 🚀

srpearce commented:
@catong This is the PR I told you about this morning. You can view @MichaelRoytman 's comments re: grammar and take a look at any other UI text as well.

@debchatigny FYI

catong commented on lms/djangoapps/support/static/support/jsx/file_upload.jsx:
```diff
@@ -0,0 +1,170 @@
+    $(e.target).val('');
+
+    if (file.size > maxFileSize) {
+      errorList.push(gettext('Please select a file with a size less than 5 MB.'));
```
@tasawernawaz  please change this message text to: "Files that you upload must be smaller than 5MB in size."

catong commented on lms/djangoapps/support/static/support/jsx/file_upload.jsx:
```diff
@@ -0,0 +1,170 @@
+    if (file.size > maxFileSize) {
+      errorList.push(gettext('Please select a file with a size less than 5 MB.'));
+    } else if ($.inArray(file.name.split('.').pop().toLowerCase(), allowedFileTypes) === -1) {
+      errorList.push(gettext('Please select an image or pdf file!'));
```
"Files that you upload must be PDFs or image files in .gif, .jpg, .jpeg, or .png format."

catong commented on lms/djangoapps/support/static/support/jsx/logged_out_user.jsx:
```diff
@@ -0,0 +1,48 @@
+    <div>
+      <div className="row">
+        <div className="col-sm-12">
+          <p>{gettext('Sign in for a faster response')}</p>
```
"Sign in to edX so we can help you better."

catong commented on lms/djangoapps/support/static/support/jsx/logged_out_user.jsx:
```diff
@@ -0,0 +1,48 @@
+      <div className="row">
+        <div className="col-sm-12">
+          <div className="form-group">
+            <label htmlFor="email">{gettext('Email')}</label>
```
"Email" -> "Your Email Address"

catong commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,211 @@
+      request.onreadystatechange = function success() {
+        if (request.readyState === 4 && request.status === 201) {
+          // TODO needs to remove after implementing success page
+          const alert = 'Request submitted successfully.';
```
@tasawernawaz Is this what the learner sees on successfully submitting the request? On the sandbox I'm seeing what might be a slightly different version where the capitalization is different. 
![supportformrequestsuccess](https://user-images.githubusercontent.com/9041399/31912311-71917386-b811-11e7-8416-4158a0699ead.png)

Could we have a more user-friendly message here?
"Thank you for submitting a support request. Someone from our Support team will contact you soon."

@dhixonedx is this OK?

tasawernawaz replied:
@catong NO. This will be replace in https://openedx.atlassian.net/browse/LEARNER-2735

catong replied:
OK! Let's remember to fix it - at the least, the message should be a complete sentence.

tasawernawaz replied:
@catong Yes we have already a story which will replace this alert.

catong commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,211 @@
+      regex = /^([a-zA-Z0-9_.+-])+@(([a-zA-Z0-9-])+\.)+([a-zA-Z0-9]{2,4})+$/;
+
+    if (!data.requester) {
+      errors.push(gettext('You must enter an email before submitting.'));
```
"Enter a valid email address."

catong commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,211 @@
+      errors.push(gettext('You must enter an email before submitting.'));
+      $('#email').closest('.form-group').addClass('has-error');
+    } else if (!regex.test(data.requester)) {
+      errors.push(gettext('You must enter a valid email before submitting.'));
```
"Enter a valid email address."

catong commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,211 @@
+      $('#email').closest('.form-group').addClass('has-error');
+    }
+    if (!data.subject) {
+      errors.push(gettext('You must enter a subject before submitting.'));
```
"Enter a subject for your support request."

catong commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,211 @@
+      $('#subject').closest('.form-group').addClass('has-error');
+    }
+    if (!data.comment.body) {
+      errors.push(gettext('You must enter a message before submitting.'));
```
"Enter some details for your support request."

catong commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,211 @@
+        <div className="row">
+          <div className="col-sm-12">
+            <div className="form-group">
+              <label htmlFor="message">{gettext('Message')}</label>
```
"Message" -> "Details"

catong commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,211 @@
+
+        <div className="row">
+          <div className="col-sm-12">
+            <p>{gettext('Your question may have already been answered.')}</p>
```
"may" -> "might"

catong commented on lms/djangoapps/support/static/support/jsx/single_support_form.jsx:
```diff
@@ -0,0 +1,211 @@
+            <a
+              href={this.props.context.marketingUrl}
+              className="btn btn-secondary help-button"
+            >{gettext('Visit edX Help')}</a>
```
"Search the edX Help Center"

tasawernawaz commented:
@catong I have addressed your comments. I have updated the sandbox. please take another look.

I believe errors should be at the top (which also was implemented in design). But yes you are right errors may not shown up for a small window, so I have added scroll to the top whenever there is any error occurred.

catong commented:
@tasawernawaz This looks great! One thing I noticed which might be unrelated to this PR or might just be the sandbox: the Sign In button at the top right seems to have black text over the dark blue background, and isn't legible. 
![2017-10-24_10-13-19](https://user-images.githubusercontent.com/9041399/31948135-0d6e42c4-b8a4-11e7-9504-7880bebdfab3.png)
